assembly/src/todo/mod.rs

```diff
@@ -0,0 +1,209 @@
+mod instruction;
+
+use core::borrow::Borrow;
+
+use vm_core::{code_blocks::CodeBlock, CodeBlockTable, Operation, Program};
+
+use crate::{
+    parsers::{self, Node, ProcedureAst, ProgramAst},
+    procedures::Procedure,
+    AssemblerError, ModuleProvider,
+};
+
+#[derive(Clone, Debug)]
+pub struct AssemblerContext {
+    cb_table: CodeBlockTable,
+    locals: Vec<Procedure>,
+}
+
+impl AssemblerContext {
+    pub fn new<I, P>(assembler: &Assembler, procs: I) -> Result<Self, AssemblerError>
+    where
+        I: Iterator<Item = P>,
+        P: Borrow<ProcedureAst>,
+    {
+        let mut context = Self {
+            cb_table: CodeBlockTable::default(),
+            locals: Vec::with_capacity(procs.size_hint().0),
+        };
+
+        for proc in procs {
+            let ProcedureAst {
+                name,
+                num_locals,
+                body,
+                is_export,
+                ..
+            } = proc.borrow();
+
+            let code_root = assembler.compile_body(&mut context, body.iter())?;
+            let proc = Procedure {
+                label: name.to_string(),
+                is_export: *is_export,
+                num_locals: *num_locals,
+                code_root,
+            };
+
+            context.locals.push(proc);
+        }
+
+        Ok(context)
+    }
+}
+
+pub struct Assembler {
+    module_provider: Box<dyn ModuleProvider>,
+}
+
+impl Assembler {
+    pub fn new() -> Self {
+        Self {
+            module_provider: Box::new(()),
+        }
+    }
+
+    pub fn with_module_provider<P>(mut self, provider: P) -> Self
+    where
+        P: ModuleProvider + 'static,
+    {
+        self.module_provider = Box::new(provider);
+        self
+    }
+
+    pub fn compile<S>(&self, source: S) -> Result<Program, AssemblerError>
+    where
+        S: AsRef<str>,
+    {
+        let source = source.as_ref();
+        let ProgramAst { local_procs, body } = parsers::parse_program(source)?;
+
+        AssemblerContext::new(self, local_procs.iter())
+            .and_then(|mut context| self.compile_body(&mut context, body.iter()))
+            .map(Program::new)
+    }
+
+    fn compile_body<A, N>(
+        &self,
+        context: &mut AssemblerContext,
+        ast: A,
+    ) -> Result<CodeBlock, AssemblerError>
+    where
+        A: Iterator<Item = N>,
+        N: Borrow<Node>,
+    {
+        let (size, hint) = ast.size_hint();
+        let size = hint.unwrap_or(size);
+
+        let mut blocks: Vec<CodeBlock> = Vec::with_capacity(size);
+        let mut span: Vec<Operation> = Vec::with_capacity(size);
+
+        for node in ast {
+            match node.borrow() {
+                Node::Instruction(instruction) => {
+                    context.compile_instruction(&mut blocks, &mut span, instruction)?;
```

Comment:
We should not put `compile_instruction()` into the `AssemblyContext` because it will create issues when parsing imported procedure calls. Instead, `compile_instruction()` should be a method on the `Assembler`.

Also, instead of passing a mutable reference to `blocks` to this function, maybe it should return an optional `CodeBlock`? For example, the method signature could be:
```Rust
fn compile_instruction(&self, instruction, &mut span) -> Result<Option<CodeBlock>, AssemblyError>
```
Then, if a `CodeBlock` is returned, we'll just push it into the `blocks` vector here. This way, when compiling `exec`, `call`, and `syscall` instructions, we'd return relevant code blocks. Otherwise, everything will work as it does now.

Reply:
Span builder added + compile instruction refactored